sdks/python/src/opik/integrations/llama_index/callback.py

```diff
@@ -0,0 +1,231 @@
+        ):
+            documents = payload_copy.pop(EventPayload.DOCUMENTS)
+            payload_copy["documents"] = [doc.metadata for doc in documents]
+            return payload_copy
+
+        if EventPayload.QUERY_STR in payload_copy:
+            return {EventPayload.QUERY_STR: payload_copy.get(EventPayload.QUERY_STR)}
+
+        if EventPayload.MESSAGES in payload_copy:
+            return {"messages": payload_copy.get(EventPayload.MESSAGES)}
+
+        if EventPayload.PROMPT in payload_copy:
+            return {"prompt": payload_copy.get(EventPayload.PROMPT)}
+
+        if payload_copy:
+            return {"input": payload_copy}
+        else:
+            return None
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: Optional[str] = None,
+        parent_id: Optional[str] = None,
+        **kwargs: Any,
+    ) -> str:
+        if not event_id:
+            event_id = str(uuid.uuid4())
+
+        # Under some scenarios, it is possible for `start_trace` to not be called (for example if
+        # the callback raises an exception in a previous call).
+        # Unclear what the best behavior is here, so for now we'll just create a new trace when
+        if self._opik_trace is None:
+            self._opik_trace = self._create_trace(trace_id=parent_id)
+
+        # Get parent span Id if it exists
+        if parent_id and parent_id in self._map_event_id_to_span:
+            opik_parent_id = self._map_event_id_to_span[parent_id].id
+        else:
+            opik_parent_id = None
+
+        # Compute the span input based on the event payload
+        span_input = self._get_span_input_from_events(event_type, payload)
+
+        # Create a new span for this event
+        span = self._opik_client.span(
+            trace_id=self._opik_trace.id,
+            name=event_type.value,
+            parent_span_id=opik_parent_id,
+            type="llm" if event_type == CBEventType.LLM else "general",
+            input=span_input,
+        )
+        self._map_event_id_to_span[event_id] = span
+
+        # If the parent_id is a BASE_TRACE_EVENT, update the trace with the span input
+        if parent_id == BASE_TRACE_EVENT and span_input:
+            self._opik_trace.update(input=span_input)
+
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: Optional[str] = None,
+        **kwargs: Any,
+    ) -> None:
+        # Get the span output from the event and store it so we can use it if needed
+        # when finishing the trace
+        span_output = self._get_span_output_from_event(event_type, payload)
+        if event_id:
+            self._map_event_id_to_output[event_id] = span_output
+
+            # Log the output to the span with the matching id
+            span = self._map_event_id_to_span[event_id]
+            if span:
```

Comment:
When is it possible? Based on the usage of `_map_event_id_to_span` - you always put not None values (spans) into dict.
And if `event_id` is not in the dict at all - the previous line will raise KeyError.

Reply:
Makes sense, updated it to not fail if the span is missing for whatever reason